analysis/analysis-api-fir/src/org/jetbrains/kotlin/analysis/api/fir/types/KtFirFunctionalType.kt

```diff
@@ -61,23 +62,28 @@ internal class KtFirFunctionalType(
             else coneType.typeArguments.size - 1
         }
 
+    private val coneTypeArgs: List<ConeKotlinType>
+        get() = coneType.valueParameterTypesIncludingReceiver(builder.rootSession)
```

Comment:
One concern is, `ownTypeArguments` have `KtType`s, and now we're building `KtType`s again for receiver, parameters, and return type. Also, this is only fixing unexpanded `typealias` in function type, but there could be more.

So, how about using `fullyExpandedType` directly at the type qualifier builder here?

https://github.com/JetBrains/kotlin/blob/master/analysis/analysis-api-fir/src/org/jetbrains/kotlin/analysis/api/fir/types/qualifiers/UsualClassTypeQualifierBuilder.kt#L31

and maybe more places of `coneType` need to be expanded as well.

Then, I guess we don't need to edit `KtFirFunctionalType` here at all.

Reply:
Done by passing expanded type to `UsualClassTypeQualifierBuilder.buildQualifiers`. The expanded type is also needed by `annotationsList`, which you mentioned in the test case, and `arity`, which is not even used anywhere and therefore not tested currently 😢.

Also I'm not sure whether expanding it for `KtFirUsualClassType` is good or not. On the other hand, expanding aliases to `FunctionN` seems fine?

Reply:
This isn't exactly what I expected, but we can revisit this for all other general cases later.